src/modules/authentication/AuthenticationWrapper.ts

```diff
@@ -0,0 +1,193 @@
+import {
+  AccountInfo,
+  AuthenticationResult, InteractionRequiredAuthError,
+  PopupRequest, SilentRequest
+} from '@azure/msal-browser';
+
+import { AUTH_URL, DEFAULT_USER_SCOPES, HOME_ACCOUNT_KEY } from '../../app/services/graph-constants';
+import { geLocale } from '../../appLocale';
+import { getCurrentUri } from './authUtils';
+import IAuthenticationWrapper from './IAuthenticationWrapper';
+import { msalApplication } from './msal-app';
+
+const defaultScopes = DEFAULT_USER_SCOPES.split(' ');
+const homeAccountKey = HOME_ACCOUNT_KEY;
+
+export class AuthenticationWrapper implements IAuthenticationWrapper {
+
+  private static instance: AuthenticationWrapper;
+  private isConsentFlow: boolean = false;
+
+  public static getInstance(): AuthenticationWrapper {
+    if (!AuthenticationWrapper.instance) {
+      AuthenticationWrapper.instance = new AuthenticationWrapper()
+    }
+    return AuthenticationWrapper.instance;
+  }
+
+  public getSessionId() {
+    const account = this.getAccount();
+    if (account) {
+      const idTokenClaims: any = account?.idTokenClaims;
+      return idTokenClaims?.sid;
+    }
+    return null;
+  }
+
+  public async logIn(sessionId = ''): Promise<AuthenticationResult> {
+    this.isConsentFlow = false;
+    try {
+      return await this.getAuthResult([], sessionId);
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  public logOut() {
+    this.deleteHomeAccountId();
+    msalApplication.logout();
+  }
+
+  public async logOutPopUp() {
+    const endSessionEndpoint = (await msalApplication.getDiscoveredAuthority()).endSessionEndpoint;
+    (window as any).open(endSessionEndpoint, 'msal', 400, 600);
+    this.clearCache();
+    this.deleteHomeAccountId();
+  }
+
+  /**
+ * Generates a new access token from passed in scopes
+ * @param {string[]} scopes passed to generate token
+ *  @returns {Promise.<AuthenticationResult>}
+ */
+  public async consentToScopes(scopes: string[] = []): Promise<AuthenticationResult> {
+    this.isConsentFlow = true;
+    try {
+      const authResult = await this.loginWithInteraction(scopes);
+      return authResult;
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  private getAccount(): AccountInfo | undefined {
+    if (!msalApplication) {
+      return undefined;
+    }
+
+    const allAccounts = msalApplication.getAllAccounts();
+    if (!allAccounts || allAccounts.length === 0) {
+      return undefined;
+    }
+
+    if (allAccounts.length > 1) {
+      const homeAccountId = this.getHomeAccountId();
+      return (homeAccountId) ? msalApplication.getAccountByHomeId(homeAccountId) || undefined : undefined;
```

Comment:
Why use msalApplication.getAccountByHomeId(homeAccountId) || undefined?

Reply:
The function `msalApplication.getAccountByHomeId(homeAccountId)` returns `AccountInfo | null`, however the references of the `getAccount()` function expects `AccountInfo | undefined`. So if the msalApp returns `null` I default it to `undefined` so that nothing breaks